keras_nlp/models/bart/bart_preprocessor.py

```diff
@@ -0,0 +1,346 @@
+from keras_nlp.models.bart.bart_presets import backbone_presets
+from keras_nlp.models.bart.bart_tokenizer import BartTokenizer
+from keras_nlp.models.preprocessor import Preprocessor
+from keras_nlp.utils.keras_utils import (
+    convert_inputs_to_list_of_tensor_segments,
+)
+from keras_nlp.utils.keras_utils import pack_x_y_sample_weight
+from keras_nlp.utils.python_utils import classproperty
+
+
+@keras_nlp_export("keras_nlp.models.BartPreprocessor")
+class BartPreprocessor(Preprocessor):
+    """A BART preprocessing layer which tokenizes and packs inputs.
+
+    This preprocessing layer will do three things:
+
+     1. Tokenize both encoder inputs and decoder inputs using the `tokenizer`.
+        Both inputs can contain any number of segments.
+     2. Pack the inputs together using `keras_nlp.layers.MultiSegmentPacker`
+        with the appropriate special tokens - `"<s>"`, `"</s>"` and `"<pad>"`.
+     3. Construct a dictionary with keys `"encoder_token_ids"`,
+        `"encoder_padding_mask"`, `"decoder_token_ids"`, `"decoder_padding_mask"`
+        that can be passed directly to a BART model.
+
+    Args:
+        tokenizer: A `keras_nlp.models.BartTokenizer` instance.
+        encoder_sequence_length: The length of the packed encoder inputs.
+        decoder_sequence_length: The length of the packed decoder inputs.
+        truncate: string. The algorithm to truncate a list of batched segments
+            to fit within `sequence_length`. The value can be either
+            `round_robin` or `waterfall`:
+                - `"round_robin"`: Available space is assigned one token at a
+                    time in a round-robin fashion to the inputs that still need
+                    some, until the limit is reached.
+                - `"waterfall"`: The allocation of the budget is done using a
+                    "waterfall" algorithm that allocates quota in a
+                    left-to-right manner and fills up the buckets until we run
+                    out of budget. It supports an arbitrary number of segments.
+
+    Call arguments:
+        x: A dictionary with `encoder_inputs` and `decoder_inputs` as its keys.
+            Each value in the dictionary can be a tensor of single string
+            sequences, or a tuple of multiple tensor sequences to be packed
+            together. Inputs may be batched or unbatched. For single sequences,
+            raw python inputs will be converted to tensors. For multiple
+            sequences, pass tensors directly.
+        y: Any label data. Will be passed through unaltered.
+        sample_weight: Any label weight data. Will be passed through unaltered.
+
+    Examples:
+
+    Directly calling the layer on data
+    ```python
+    preprocessor = keras_nlp.models.BartPreprocessor.from_preset("bart_base_en")
+
+    # Tokenize and pack a single sentence.
+    inputs = {
+        "encoder_inputs": "The fox was sleeping.",
```

Comment:
Open question...

Should we call this `"encoder_text"` to better accommodate `"encoder_audio"` for whisper? Or will it be simpler to have the same names everywhere. I somewhat like the self documenting property of saying this is text input.

Reply:
Yep, `"encoder_text"` sounds good to me!